packages/abi-coder/src/types/JsonAbiNew.ts

```diff
@@ -0,0 +1,101 @@
+/**
+ * Types for Fuel JSON ABI Format as defined on:
+ * https://github.com/FuelLabs/fuel-specs/blob/master/src/abi/json-abi-format.md
+ */
+export interface JsonAbi {
+  readonly specVersion: string;
+  readonly encodingVersion: string;
+  readonly programType: string;
+  readonly concreteTypes: readonly ConcreteType[];
+  readonly metadataTypes: readonly MetadataType[];
+  readonly functions: readonly AbiFunction[];
+  readonly loggedTypes: readonly LoggedType[];
+  readonly messagesTypes: readonly MessageType[];
+  readonly configurables: readonly Configurable[];
+}
+
+export interface ConcreteType {
+  readonly type: string;
+  readonly concreteTypeId: string;
+  readonly metadataTypeId?: number;
+  readonly typeArguments?: readonly string[];
+}
+
+export interface MetadataType {
+  readonly type: string;
+  readonly metadataTypeId: number;
+  readonly components?: readonly Component[];
+  readonly typeParameters?: readonly number[];
+}
+
+export interface Component extends TypeArgument {
+  readonly name: string;
+}
+
+export interface TypeArgument {
+  readonly typeId: number | string; // the type metadata declaration ID or type concrete declaration hash based ID of the type of the component.
```

Comment:
How do we know which array to look for the `typeId` in? If it's a number, it's the `metadataTypes` array?

Reply:
Yes, if it's a number then it's in the `metadataTypes` array, and if it's a string then it's in the `concreteTypes` array.

Are you doing some parsing of the abi on your own in some applications hence why you're asking the questions? If yes, I'd like to see where it's done and why because ideally the `Interface` should be sufficient for all abi-related requirements.

Reply:
Sorry I missed this. Yes I have to do all of this in my application. See [here](https://github.com/FuelLabs/sway-playground/blob/6ae1b33ce1e4bcf589f2acfa25fb89c827b8e5ac/app/src/features/interact/components/ContractInterface.tsx#L27)

I'm curious why this is the hash for concrete types rather than the array index.

This new interface in the SDK is pretty complex.

Reply:
> Yes I have to do all of this in my application. See [here](https://github.com/FuelLabs/sway-playground/blob/6ae1b33ce1e4bcf589f2acfa25fb89c827b8e5ac/app/src/features/interact/components/ContractInterface.tsx#L27)

Okay, so you have to parse the abi to present the infinite amount of possible inputs the users can generate. The SDK doesn't have something to support this as it's a highly specific use case that we haven't encountered previously because people are creating apps for already known sway programs and they know their inputs. 

> I'm curious why this is the hash for concrete types rather than the array index.

Because of the need for deterministic log ids. Previously array indexes would change and somebody doing e.g. `decodeLog(56, value)` would get a failure if the array changed. Now they pass in a hash of the concrete type and it won't change when the abi changes if the abi change doesn't affect that type.

> This new interface in the SDK is much more complex. It would have been nice to preserve the old interface or at least provide helper functions for conversion.

It just follows the specs. We intentionally wanted to remove the old interface so that we can stop supporting it altogether as `forc 0.63.0` and upwards won't be producing the old abi format. 

You _could_ internalize the [transpiler](https://github.com/FuelLabs/fuels-ts/pull/2856/files#diff-84ad7158971cfbd3205763a6ca8cf5d047ee10833b29074ecc5e48cc3fbdc33dR87) from this PR in your application, as it's doing the job of converting the new abi into the old, as well as the [old `JsonAbi` interface](https://github.com/FuelLabs/fuels-ts/pull/2856/files#diff-57ca93f87291b2c564b725c691f0f4e100de69a659a9b0eda64786757bdbad00R5).

We will be removing the tranpiler as part of #2877 because it's just there to give us more room to think about refactoring the packages that interact with `JsonAbi`.

Reply:
Will the parse functions like `parseComponents` and `parseConcreteTypes` also be removed? Or is this something I can use?

Reply:
Both of those functions are internal to the SDK and you aren't able to access them as they aren't exported. Everything introduced in this PR in relation to transpiling will be removed - those functions included. You can only go to the source code and copy/paste the whole logic for yourself and it'll be something that you'll have to maintain in your apps.